### PR TITLE
Add step_args parameter to LRScheduler.simulate

### PR DESCRIPTION
**Pull request by githubnemo**
The purpose of this change is to allow for simulation of scheduling policies such as `ReduceLROnPlateau` to be simulated properly.

If `step_args` is an indexable object, it is indexed using the current simulated epoch to get closer to real-life behavior, e.g. simulating a real loss curve and the corresponding behavior of the LR scheduler policy.

### CODE REVIEW

File: skorch/tests/callbacks/test_lr_scheduler.py [hunk 1]
```diff
@@ -35,6 +35,41 @@ def test_simulate_lrs_batch_step(self, policy):
...
+    def test_simulate_lrs_reduced_lr_on_plateau_scalar(self):
+        # Feed a constant, scalar "loss" to the scheduler.
+        lr_sch = LRScheduler(
+            ReduceLROnPlateau, factor=0.1, patience=1,
+        )
+        lrs = lr_sch.simulate(
+            steps=5, initial_lr=1, step_args=0.5
+        )
+        # O = OK epoch
+        # I = intertolerable epoch
```
Comment (BenjaminBossan) [inline skorch/tests/callbacks/test_lr_scheduler.py:47]: What is "intertolerable"?

File: skorch/tests/callbacks/test_lr_scheduler.py [hunk 2]
```diff
@@ -35,6 +35,41 @@ def test_simulate_lrs_batch_step(self, policy):
...
+        #
+        # O I I I I   epoch classification
+        # 0 1 2 3 4   number of bad epochs
+        #     * * *   epochs with LR reduction
```
Comment (BenjaminBossan) [inline skorch/tests/callbacks/test_lr_scheduler.py:51]: Should it be:
```suggestion
        # O I I I I   epoch classification
        # 0 1 2 3 4   number of bad epochs
        #     *       epochs with LR reduction
```

File: skorch/tests/callbacks/test_lr_scheduler.py [hunk 3]
```diff
@@ -35,6 +35,41 @@ def test_simulate_lrs_batch_step(self, policy):
...
+        #
+        # note that simulate returns the LRs before the step, not after,
+        # so we're seeing only 4 updated values.
+        assert all(lrs == [1] + [1, 1, 0.1, 0.1])
```
Comment (BenjaminBossan) [inline skorch/tests/callbacks/test_lr_scheduler.py:55]: Do we see "4 updated values"? Why use `[1] + ...` here but not in the next test?

File: skorch/callbacks/lr_scheduler.py [hunk 4]
```diff
@@ -87,6 +87,11 @@ def simulate(self, steps, initial_lr):
         initial_lr: float
           Initial learning rate
 
+        step_args: Any
```
Comment (BenjaminBossan) [inline skorch/callbacks/lr_scheduler.py:90]: Would it be more helpful to change the type to `None or float or list of float (default=None)`?

Yes, technically it doesn't have to be a list but I don't think it still makes it easier for users to know what they need to pass.

File: skorch/callbacks/lr_scheduler.py [hunk 5]
```diff
@@ -87,6 +87,11 @@ def simulate(self, steps, initial_lr):
...
+          Argument to the `.step()` function of the policy. If it is an
+          indexable object the simulation will try to associate every step of
+          the simulation with an entry in ``step_args``.
```
Comment (BenjaminBossan) [inline skorch/callbacks/lr_scheduler.py:93]: Also describe what happens when `None` or `float` are passed?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the PR to enhance the capabilities of `simulate`. My comments are mostly about clarifications.

Failing tests require #1078 to be merged.

File: skorch/tests/callbacks/test_lr_scheduler.py (hunk 1 above)
Comment (githubnemo) [reply skorch/tests/callbacks/test_lr_scheduler.py:47]: This is terminology taken from the [docs](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html#torch.optim.lr_scheduler.ReduceLROnPlateau)

> In the second epoch, if the performance is worse than the baseline, we have what is considered an intolerable epoch.

Drastic terminology but I wanted to adhere to already defined terms :)

Comment (BenjaminBossan) [reply skorch/tests/callbacks/test_lr_scheduler.py:47]: Drastic indeed, but it's in the Shakespeare corpus, so it must be a common English word.

File: skorch/tests/callbacks/test_lr_scheduler.py (hunk 2 above)
Comment (githubnemo) [reply skorch/tests/callbacks/test_lr_scheduler.py]: No, it's actually

```python
        # O I I I I   epoch classification
        # 0 1 2 1 2   number of bad epochs
        #     *   *   epochs with LR reduction
```

I've updated the comment but the gist is that the bad epoch counter is reset after an LR reduction.

File: skorch/tests/callbacks/test_lr_scheduler.py (hunk 3 above)
Comment (githubnemo) [reply skorch/tests/callbacks/test_lr_scheduler.py]: I've removed the `[1]`, it did not help visualize what's going on. I've also changed it to "4 simulated values" since the first value to return is always the init. LR.

File: skorch/callbacks/lr_scheduler.py (hunk 4 above)
Comment (githubnemo) [reply skorch/callbacks/lr_scheduler.py]: Agreed. Changed!

**Review by BenjaminBossan — APPROVED** [review]
Thanks for the changes, it's now much easier to understand what happens.
